tower/src/util/combinators/map_err.rs

```diff
@@ -0,0 +1,70 @@
+use futures_util::{future::MapErr as MapErrFut, TryFutureExt};
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Service returned by the [`map_err`] combinator.
+///
+/// [`map_err`]: crate::util::ServiceExt::map_err
+#[derive(Debug)]
+pub struct MapErr<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> MapErr<S, F> {
+    /// Creates a new [`MapErr`] service.
+    pub fn new(inner: S, f: F) -> Self {
+        MapErr { f, inner }
+    }
+}
+
+impl<S, F, Request, Error> Service<Request> for MapErr<S, F>
+where
+    S: Service<Request>,
+    F: FnOnce(S::Error) -> Error + Clone,
```

Comment:
Why clone? Why not `Fn(S::Error) -> Error`? I believe `Fn` is semantically similar to `FnOnce + Clone` while being less restrictive.

Reply:
@LucioFranco I don't think that's the case. `Fn` can be called with an `&` reference. However, since the function is called _after_ the inner service's future, the function needs to be moved into the future that the `MapErr` service returns. Therefore, it cannot be borrowed without constraining the future's lifetime --- so it must be `Clone`d.  And if it _has_ to be `Clone`, it may as well be `FnOnce`, since each clone will only be called a single time.

If we wanted to use `Fn`, we would need to wrap it in an `Arc` internally, and clone the `Arc` into each future. That would also work, but in some cases (such as function pointers), the `Arc` isn't needed, since little-f `fn` can be freely cloned already.

Reply:
@hawkw is right, we should then change the bounds on the `map` combinators as well.